### PR TITLE
Image v3 Auto Dynamic Media Not Working

### LINKED ISSUE #2042: Image v3 Auto Dynamic Media Not Working
## Bug Report

**Current Behavior**
When using the "Auto" rendition option with the Image v3 component, the auto rendition option is not updating the image source url. 

**Expected behavior/code**
My understanding and experience using the v2 image component was that you needed to utilize Javascript to get the Dynamic Media auto option working. With the v3 version of the component it doesn't mention how the DM is supposed to work. 

**Environment**
- AEM version and patch level (e.g. AEM 6.5.10 SP1 CFP2)
- Core Components version (e.g. 2.18.0)
- JRE version: 11.0.12

**Additional context / Screenshots**
Screen Shot 2022-02-28 at 9 13 55 PM

Screen Shot 2022-02-28 at 9 32 16 PM

Screen Shot 2022-02-28 at 9 34 59 PM

### LINKED ISSUE #2140: [Imagev3] Add url encoding to DM image srcset attribute
## Bug Report

**Current Behavior**
srcset is built using the output of the method  `getSrcUriTemplate()` and `getWidths()`.
In order to build it, srcUriTemplate is decoded and manipulated replacing the {width}, however it is not encoded again before returned.
This causes problems for image urls that require encoding, for instance because they have blank spaces in the url.
E.g.
a srcUriTemplate like: `https://assets.georgfischer.com/is/image/georgfischeragstage/Bolivia%20Project_Hero?qlt=82&ts=1644856523736&dpr=off`
generates srcset like: `https://assets.georgfischer.com/is/image/georgfischeragstage/Bolivia Project_Hero?qlt=82&amp;wid=1280&amp;ts=1644856523736&amp;dpr=off 1280w`

Notice the %20 missing in the image name "Bolivia Project_Hero"

**Expected behavior/code**
image urls in srcset should be encoded

**Environment**
- AEM Cloud Service 2022.3.6698.20220318T233218Z

**Possible Solution**
Encode the urls in srcset before returning them

### PR DESCRIPTION
**Pull request by bpauli**
- added client library to set sizes attribute automatically and to set correct srcset attribute for dm images while using smart crop auto feature
- add required data attributes to imageV3 markup
- set empty srcset in sling model for dm images while using smart crop auto feature
- adapt unit test

fixes #2042 #2140

| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #2042, Fixes #2140
| Patch: Bug Fix?          | 👍
| Minor: New Feature?      | 👎
| Major: Breaking Change?  | 👎
| Tests Added + Pass?      | Yes
| Documentation Provided   | Yes (code comments and or markdown)
| Any Dependency Changes?  |
| License                  | Apache License, Version 2.0

### CODE REVIEW

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v3/ImageImpl.java [hunk 1]
```diff
@@ -108,22 +108,22 @@ public String getSrcset() {
...
         String[] srcsetArray = new String[widthsArray.length];
         if (widthsArray.length > 0 && srcUritemplate != null) {
-            String srcUriTemplateDecoded = "";
-            try {
-                srcUriTemplateDecoded = URLDecoder.decode(srcUritemplate, StandardCharsets.UTF_8.name());
-            } catch (UnsupportedEncodingException e) {
-                LOGGER.error("Character Decoding failed for " + resource.getPath());
-            }
-            if (srcUriTemplateDecoded.contains("{.width}")) {
+            srcUritemplate = StringUtils.replace(srcUriTemplate, URI_WIDTH_PLACEHOLDER_ENCODED, URI_WIDTH_PLACEHOLDER);
```
Comment (vladbailescu) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v3/ImageImpl.java:120]: Not sure why you're skipping decoding the `srcUriTemplate` now, won't this lead to double encoding of `srcsetArray` items?

File: bundles/core/src/test/resources/image/v3/exporter-image38-two-smart-sizes.json [hunk 2]
```diff
@@ -2,8 +2,7 @@
   "id": "image-cfa1cb817f",
   "alt": "Adobe Systems Logo and Wordmark in PNG format",
   "title": "Adobe Systems Logo and Wordmark",
-  "srcset":"https://s7d9.scene7.com/is/image/dmtestcompany/Adobe_Systems_logo_and_wordmark_DM.600?ts=1490005239000&dpr=off 600w,https://s7d9.scene7.com/is/image/dmtestcompany/Adobe_Systems_logo_and_wordmark_DM.800?ts=1490005239000&dpr=off 800w",
-  "src": "https://s7d9.scene7.com/is/image/dmtestcompany/Adobe_Systems_logo_and_wordmark_DM?ts=1490005239000&dpr=off",
+  "src": "https://s7d9.scene7.com/is/image/dmtestcompany/Adobe_Systems_logo_and_wordmark_DM?ts=1490005239000&dpr=on,{dpr}",
```
Comment (vladbailescu) [inline bundles/core/src/test/resources/image/v3/exporter-image38-two-smart-sizes.json:5]: Shouldn't the template also be part of the exported JSON? How would a clientside/SPA implementation rebuild the component?

File: bundles/core/src/test/resources/image/v3/exporter-image38.json [hunk 3]
```diff
@@ -1,28 +1,28 @@
...
-      "xdm:linkURL": "https://www.adobe.com",
-      "image": {
-        "repo:id": "60a1a56e-f3f4-4021-a7bf-ac7a51f0ffe5",
-        "repo:modifyDate": "2017-03-20T10:20:39Z",
-        "@type": "image/png",
-        "repo:path": "/content/dam/core/images/Adobe_Systems_logo_and_wordmark_DM.png"
-      }
-    }
-  }
+    ":type": "core/wcm/components/image/v3/image",
```
Comment (vladbailescu) [inline bundles/core/src/test/resources/image/v3/exporter-image38.json:2]: Hard to tell what changed (if anything)

File: bundles/core/src/test/resources/image/v3/exporter-image39.json [hunk 4]
```diff
@@ -1,28 +1,28 @@
...
-      "xdm:linkURL": "https://www.adobe.com",
-      "image": {
-        "repo:id": "60a1a56e-f3f4-4021-a7bf-ac7a51f0ffe5",
-        "repo:modifyDate": "2017-03-20T10:20:39Z",
-        "@type": "image/png",
-        "repo:path": "/content/dam/core/images/Adobe_Systems_logo_and_wordmark_DM.png"
-      }
-    }
-  }
+    ":type": "core/wcm/components/image/v3/image",
```
Comment (vladbailescu) [inline bundles/core/src/test/resources/image/v3/exporter-image39.json:2]: Hard to tell what changed (if anything)

File: content/src/content/jcr_root/apps/core/wcm/components/image/v3/image/clientlibs/site/js/image.js [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        event.initEvent(name, true, true);
+        event.detail = detail;
+        elem.dispatchEvent(event);
+        return event;
+    };
+
+    var debounce = function(func) {
+        var timeout;
+        var timestamp;
+        var wait = 99;
```
Comment (vladbailescu) [inline content/src/content/jcr_root/apps/core/wcm/components/image/v3/image/clientlibs/site/js/image.js:84]: We should probably define this in a more prominent location (as a constant).

File: content/src/content/jcr_root/apps/core/wcm/components/image/v3/image/clientlibs/site/js/image.js [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+        };
+        request.send();
+        return autoSmartCrops;
+    };
+
+    var image = (function() {
+        var images;
+
+        var setSizes = function(imgEl, width) {
+            imgEl._autoWidth = width;
```
Comment (vladbailescu) [inline content/src/content/jcr_root/apps/core/wcm/components/image/v3/image/clientlibs/site/js/image.js:148]: This code needs some documentation, it's hard to follow the logic. Also, is it even safe to use `_autoWidth`?

File: content/src/content/jcr_root/apps/core/wcm/components/image/v3/image/clientlibs/site/js/image.js [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+        return function() {
+            timestamp = Date.now();
+
+            if (!timeout) {
+                timeout = setTimeout(later, wait);
+            }
+        };
+    };
+
+    var getAutoSmartCrops = function(src) {
```
Comment (vladbailescu) [inline content/src/content/jcr_root/apps/core/wcm/components/image/v3/image/clientlibs/site/js/image.js:108]: I would move this utility method to a common clientlib

**Review by vladbailescu — CHANGES_REQUESTED** [review]
I added a couple of suggestions on things that could be improved.

We also need to figure out a way to test the behaviour.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v3/ImageImpl.java (hunk 1 above)
Comment (bpauli) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v3/ImageImpl.java:120]: I removed the decoding to fix the issue https://github.com/adobe/aem-core-wcm-components/issues/2140 and to keep the encoding (e.g for whitespaces). Why do think there will be a double encoding?

File: bundles/core/src/test/resources/image/v3/exporter-image38-two-smart-sizes.json (hunk 2 above)
Comment (bpauli) [reply bundles/core/src/test/resources/image/v3/exporter-image38-two-smart-sizes.json:5]: The current exported value for the `srcset` is wrong for dm images which have a `smartCropRendition=SmartCrop:Auto` and should be considered as a 🐛. In the other PR I tried to generate the array on server side but this approach was not successful because:
1. The `DynamicMediaRenditionProvider` is not reliable. Sometimes the renditions got not created in the JCR.
2. The device pixel ratio can only be detected on client side.

For a clientide/SPA implementation it should be easy to adapt the existing client side solution of this PR.

**Comment by yugandhar02** [comment]
Is there any setup where i can test the fix ?

**Comment by bpauli** [comment]
close this one in favour of https://github.com/adobe/aem-core-wcm-components/pull/2249
